src/python/pants/backend/go/target_type_rules.py

```diff
@@ -213,6 +228,58 @@ async def inject_go_external_package_dependencies(
     return InjectedDependencies(inferred_dependencies)
 
 
+# -----------------------------------------------------------------------------------------------
+# Generate `_go_external_package` targets
+# -----------------------------------------------------------------------------------------------
+
+
+class GenerateGoExternalPackageTargetsRequest(GenerateTargetsRequest):
+    generate_from = GoModule
+
+
+@rule(desc="Generate targets for each external package in `go.mod`", level=LogLevel.DEBUG)
+async def generate_go_external_package_targets(
+    request: GenerateGoExternalPackageTargetsRequest,
+) -> GeneratedTargets:
+    generator_addr = request.generator.address
+    resolved_module = await Get(ResolvedGoModule, ResolveGoModuleRequest(generator_addr))
```

Comment:
If you comment out the below MultiGet, this still causes a cycle. It is fixed by replacing this line:

https://github.com/pantsbuild/pants/blob/68a76e031b0d46f67f71bcec37a9f41a96ce2cca/src/python/pants/backend/go/module.py#L99

with a manual `Get(Snapshot, PathGlobs)`, which basically does what `HydrateSources` does but disables all the codegen stuff.